pkg/fqdn/name_manager.go

```diff
@@ -139,11 +145,21 @@ func NewNameManager(config Config) *NameManager {
 		}
 	}
 
+	// Setup the update queue and a trivial goroutine to apply them
+	updateQueue := make(chan func(), 1024)
+	go func() {
+		defer log.Warn("Unexpected exit of NameManager update goroutine. FQDN -> Policy updates will stop working until cilium-agent is restarted")
+		for update := range updateQueue {
```

Comment:
Does Golang guarantee that this goroutine won't somehow be scheduled from the same thread as the goroutine that inserts into the `updateQueue`? Because that would presumably still allow for the deadlock to be hit.

Reply:
AFAIK a goroutine can't be blocked due to another goroutine being scheduled in the same "thread". It seems to me a deadlock is possible only if the `update` function take the name manager lock (which I don't think is the case).

Reply:
From the OS perspective, no, but all the go synchronisation primitives would yield between goroutines.

Reply:
My concern is that the goroutine that is executing `UpdateGenerateDNS()` -> `n.enqueueUpdate()` is holding the name-manager lock, then yields to this goroutine, which tries to execute `update()` which grabs the SelectorCache lock.

Then at the same time we have the other path from before from another separate goroutine which grabs these two locks in the reverse order:

```
pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock { i.Mutex.Lock() } <<<<<
pkg/fqdn/name_manager.go:120 fqdn.(*NameManager).UnregisterForIdentityUpdates { n.Mutex.Lock() }
pkg/policy/selectorcache.go:302 policy.(*fqdnSelector).removeUser { f.dnsProxy.UnregisterForIdentityUpdates(f.selector) }
pkg/policy/selectorcache.go:677 policy.(*SelectorCache).removeSelectorLocked { if sel.removeUser(user) { }
pkg/policy/selectorcache.go:694 policy.(*SelectorCache).RemoveSelectors { sc.removeSelectorLocked(selector, user) }
pkg/policy/l4.go:361 policy.(*L4Filter).detach { selectorCache.RemoveSelectors(l4.CachedSelectors, l4) }
pkg/policy/l4.go:466 policy.L4PolicyMap.Detach { f.detach(selectorCache) }
pkg/policy/l4.go:625 policy.(*L4Policy).Detach { l4.Egress.Detach(selectorCache) }
pkg/policy/resolve.go:104 policy.(*selectorPolicy).Detach { p.L4Policy.Detach(p.SelectorCache) }
pkg/policy/resolve.go:103 policy.(*selectorPolicy).Detach { if p.L4Policy != nil { }
pkg/policy/distillery.go:140 policy.(*PolicyCache).updateSelectorPolicy { cip.setPolicy(selPolicy) }
pkg/policy/distillery.go:170 policy.(*PolicyCache).UpdatePolicy { _, err := cache.updateSelectorPolicy(identity) }
pkg/endpoint/policy.go:169 endpoint.(*Endpoint).regeneratePolicy { if err := repo.GetPolicyCache().UpdatePolicy(e.SecurityIdentity); err != nil { }
```

Reply:
Conceptually the yielding ends when the receiving goroutine reads from the channel, so they will be executing in parallel when the `update` is executing. Also, as Ray said, it seems to me that the lock ordering analysis does not span goroutines in any case.

Reply:
OK, so the `update()` goroutine would potentially yield when attempting to grab the mutex, then the `UpdateGenerateDNS()` goroutine is released and can unlock the mutex to allow other acquirers of the `NameManager` mutex to continue.

Then if there's other `UpdateGenerateDNS()` goroutines waiting in the wings for the NM mutex, they will attempt the non-blocking send and fail out with a log, avoiding the deadlock.